rdt/hyper_transformer.py

```
@@ -17,136 +10,163 @@ class HyperTransformer:
+            Dict mapping data types to transformers to use for that data type.
         copy (bool):
             Whether to make a copy of the input data or not. Defaults to ``True``.
-        dtypes (list or None):
-            List of column data types to use when building the ``transformers`` dict
-            automatically. If not passed, the ``DataFrame.dtypes`` are used.
-        dtype_transformers (dict or None):
-            Transformer templates to use for each dtype. Passed as a dictionary of
-            dtype kinds ('i', 'f', 'O', 'b', 'M') and transformer names, classes
-            or instances.
+        transform_output_types (list or None):
+            List of acceptable data types for the output of the ``transform`` method.
+            If ``None``, only ``numerical`` types will be considered acceptable.
+
 
     Example:
         Create a simple ``HyperTransformer`` instance that will decide which transformers
         to use based on the fit data ``dtypes``.
 
         >>> ht = HyperTransformer()
 
-        Create a ``HyperTransformer`` passing a list of dtypes.
+        Create a ``HyperTransformer`` passing a dict mapping fields to data types.
 
-        >>> ht = HyperTransformer(dtypes=[int, 'object', np.float64, 'datetime', 'bool'])
+        >>> field_types = {
+        ...     'a': 'categorical',
+        ...     'b': 'numerical
+        ... }
+        >>> ht = HyperTransformer(field_types=field_types)
+
+        Create a ``HyperTransformer`` passing a ``field_transformers`` dict.
+        (Note: The transformers used in this example may not exist and are just used
+        to illustrate the different way that a transformer can be defined for a field).
 
-        Create a ``HyperTransformer`` passing a ``transformers`` dict.
+        >>> field_transformers = {
+        ...     'email': EmailTransformer(),
+        ...     'email.domain': EmailDomainTransformer(),
+        ...     ('year', 'month', 'day'): DateTimeTransformer()
```

I have a question: what's the expected behavior if a field is specified multiple times? For example:

```
field_transformers = {
    'email': EmailTransformer(),
    ('email',): OneHotEncoding(),
    ('email', 'phone_number'): OneHotEncoding(),
}
```
If this is not allowed, maybe we should specify it?

---

Good point Felipe. I think that at some point we discussed that this would be validated, so the same column name can only be mentioned once, but it seems at the end it was not added.

---

Maybe I'll add a warning whenever there are multiple definitions.

---

Actually I don't think this makes sense. A dictionary can't have the same key twice. If they specify a tuple with only one value, that is a nonsensical input. I don't know if that's worth validating

---

I am thinking that this should be an error that makes the code crash rather than a warning, because if the user really passes the same field twice we cannot work around it